src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -1,64 +1,179 @@
+        }
+    }
+
+    fun buildCoroutineGraph(suspendContext: SuspendContext) {
+        val suspendContextImpl = suspendContext as? SuspendContextImpl ?: run {
+            emptyText.text = CoroutineStacksBundle.message("coroutine.stacks.could.not.be.built")
+            return
+        }
+
+        val coroutineInfoCache = CoroutineDebugProbesProxy(suspendContextImpl).dumpCoroutines()
+        val coroutineInfoDataList = coroutineInfoCache.cache
+        val dispatcherToCoroutineDataList = mutableMapOf<String, MutableList<CoroutineInfoData>>()
+        for (info in coroutineInfoDataList) {
+            val dispatcher = info.descriptor.dispatcher ?: continue
+            dispatcherToCoroutineDataList.computeIfAbsent(dispatcher) { mutableListOf() }.add(info)
+        }
+
+        val dispatcherLabel = JLabel(CoroutineStacksBundle.message("select.dispatcher"))
+        val coroutineStacksWindowHeader = Box.createHorizontalBox()
+        val dispatchers = dispatcherToCoroutineDataList.keys.toTypedArray()
+        val dispatcherDropdownMenu = ComboBox(dispatchers)
+
+        dispatcherDropdownMenu.addActionListener {
+            val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
+            val coroutineDataList = dispatcherToCoroutineDataList[selectedDispatcher]
+            if (!coroutineDataList.isNullOrEmpty()) {
+                updateCoroutineStackForest(coroutineDataList, suspendContextImpl)
+            }
+        }
+
+        val comboBoxSize = Dimension(Constants.comboBoxHeight, Constants.comboBoxWidth)
+        dispatcherDropdownMenu.apply {
+            preferredSize = comboBoxSize
+            maximumSize = comboBoxSize
+            minimumSize = comboBoxSize
+        }
+        coroutineStacksWindowHeader.add(dispatcherLabel)
+        coroutineStacksWindowHeader.add(dispatcherDropdownMenu)
+
+        dispatcherToCoroutineDataList.forEach { (dispatcher, coroutineDataList) ->
+            coroutineDataList.forEach { data ->
+                if (data.descriptor.state == State.RUNNING) {
+                    dispatcherDropdownMenu.selectedItem = dispatcher
+                }
             }
         }
 
-        coroutineGraph = mxGraphComponent(graph)
-        add(coroutineGraph)
+        val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
+        val coroutineDataList = dispatcherToCoroutineDataList[selectedDispatcher]
+        if (!coroutineDataList.isNullOrEmpty()) {
+            updateCoroutineStackForest(coroutineDataList, suspendContextImpl)
+        }
+        panelContent.add(coroutineStacksWindowHeader)
+        panelContent.add(forest)
+        add(panelContent)
+    }
+
+    private fun updateCoroutineStackForest(
+        coroutineDataList: MutableList<CoroutineInfoData>,
+        suspendContextImpl: SuspendContextImpl
+    ) {
+        forest.removeAll()
+        val root = Node()
+        val coroutineStackForest = suspendContextImpl.buildCoroutineStackForest(
```

Comment:
We need to wrap the stack forest building logic in the debugger thread, and the UI building logic in the UI thread